### PR TITLE
Add guardrail to detect non-uniform sampling intervals (#300)

### LINKED ISSUE #300: Add guardrail for hardware-specific quirks: detect non-uniform sampling intervals
## Summary
Different eye trackers have hardware-specific quirks that the current pipeline's parameters and logic implicitly assume away (EyeLink-specific behavior). When supporting other trackers, we need guardrails to catch these.

## Background
Reviewer noted that supporting new trackers isn't just about loaders — pipeline parameters and logic depend on tracker quirks. Two examples raised:
- EyeLink tends to produce sharp downward spikes in pupil data around blinks (not all trackers do this)
- Some hardware **drops samples** instead of filling with `0` when pupil data is missing

## Proposed solution
Add a guardrail function that checks for **consecutive, equal time intervals** in the input timeseries to catch dropped-sample quirks early.

## Acceptance criteria
- [ ] Function validates consecutive sampling intervals are equal/uniform
- [ ] Emits informative warning/error when irregular intervals detected
- [ ] Integrated into the generic loader / pipeline entry
- [ ] Test coverage for the dropped-sample case

## Notes
We have observed these quirks with other hardware in the past, including dropped samples vs. zero-fill on missing pupil data.

### PR DESCRIPTION
**Pull request by shawntz**
Closes #300.

## Summary

Different eye trackers have hardware-specific quirks that violate the uniform-sampling assumption baked into the eyeris pipeline. Notably, some hardware **drops samples** when pupil data is missing instead of zero-filling (as EyeLink does), leaving gaps in the otherwise evenly spaced time vector. Because downstream steps (`detransient()`, `lpfilt()`, `downsample()`) assume a fixed sampling rate, such gaps can silently distort results.

This PR adds a guardrail that surfaces the quirk early, at load time.

## What changed

- **`R/utils-checks.R`** — new internal `check_uniform_sampling_intervals()` (+ `modal_value()` helper):
  - Infers the expected inter-sample interval from the data's **modal positive interval** — robust to dropped samples *and* to sub-millisecond timestamp rounding at high sampling rates.
  - **(A) Sporadic dropout:** flags intervals exceeding `expected × (1 + tolerance)`, estimates the number of dropped samples, and reports the largest gap.
  - **(B) Systematic dropout:** cross-checks the data-derived interval against the device's reported `hz` to catch the case where *every Nth sample* is missing — which leaves a uniform-but-coarser grid that gap detection alone cannot see. Guarded against false positives on high-rate trackers that report integer-millisecond timestamps for sub-millisecond samples (uses zero-length intervals as the tell-tale).
  - Checks each **recording segment (`block`) independently**, so legitimate between-block gaps are not flagged.
  - Emits informative `log_warn` messages and returns a structured summary invisibly.
- **`R/pipeline-loadasc.R`** — wired into `load_asc()` once, on the shared raw `time`/`block`, **before** the binocular split (covers monocular and all binocular modes with no duplicate warnings).
- **`tests/testthat/test-uniform-sampling.R`** — 15 tests / 56 assertions.
- **`NEWS.md`** entry + regenerated man pages.

## Acceptance criteria

- [x] Function validates consecutive sampling intervals are equal/uniform
- [x] Emits informative warning when irregular intervals detected
- [x] Integrated into the loader / pipeline entry (`load_asc()`)
- [x] Test coverage for the dropped-sample case (sporadic + systematic)

## Verification

- New tests **56/56 pass**; full suite **293 pass**. The only 4 failures are **pre-existing** environmental flakes in `test-parallel-database.R` (fail identically on the clean tree — unrelated to this change).
- Both bundled demo datasets load with **zero** false warnings across monocular + all four binocular modes.
- `spell_check_package()` clean; `NAMESPACE` unchanged (functions are internal).

## Notes

- The new functions are `@keywords internal` (not exported), matching the existing `check_*` guardrails like `check_time_monotonic()`.
- The NEWS entry is filed under a new `# eyeris (development version)` heading; `DESCRIPTION` `Version` is intentionally left untouched for you to bump at release time.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added validation during ASC data loading to detect irregular or missing sampling intervals.
  * Provides warnings for non-uniform timestamps and systematic sampling-rate mismatches.
  * Supports checking separate recording segments and suppressing warnings when needed.
  * Added documentation describing sampling-interval validation and interpretation.

* **Bug Fixes**
  * Prevents gaps at recording-block boundaries from being incorrectly reported as dropped samples.

* **Tests**
  * Added comprehensive coverage for uniform sampling, irregular intervals, missing samples, rate mismatches, and edge cases.

### CODE REVIEW

File: R/utils-checks.R [hunk 1]
```diff
@@ -225,6 +225,234 @@ check_column <- function(df, col_name) {
...
+  expected <- modal_value(positive_intervals)
+  if (!is.finite(expected) || expected <= 0) {
+    expected <- if (!is.null(hz) && is.finite(hz) && hz > 0) {
+      1000 / hz
+    } else {
+      NA_real_
+    }
+  }
+  if (!is.finite(expected) || expected <= 0) {
+    return(invisible(result))
```
Comment (mh105) [inline R/utils-checks.R]: I think this fallback is too defensive for a guardrail. If we cannot infer a finite positive interval from the timestamp differences, the timestamp vector itself is not suitable for this check; falling back to `1000 / hz` could hide malformed timestamps and return a misleading result. Could we fail loudly instead?

```suggestion
  if (!is.finite(expected) || expected <= 0) {
    log_error(
      "Unable to infer expected sampling interval from timestamp differences. ",
      "Check that sample timestamps are finite and increasing."
    )
  }
```

**Review by mh105 — COMMENTED** [review]
Suggested change for the interval inference fallback.

File: R/utils-checks.R [hunk 2]
```diff
@@ -225,6 +225,234 @@ check_column <- function(df, col_name) {
...
+  }
+
+  result$expected_interval <- expected
+  n_total <- length(intervals)
+  result$n_intervals <- n_total
+
+  # (A) gap detection: intervals materially larger than the expected spacing
+  # (i.e., sporadic dropped samples, leaving holes in the time grid)
+  irregular <- intervals > expected * (1 + tolerance)
```
Comment (mh105) [inline R/utils-checks.R]: This currently behaves like a dropped-sample detector rather than a uniform-grid check: it only flags intervals larger than expected. If downstream signal processing assumes a fixed sampling interval, shorter intervals are also a violation. For example, 250 Hz data should be exactly 4 ms apart, so both `c(4, 5, 4)` and `c(4, 3, 4)` should be reported as irregular; the regularizer can decide how to repair those cases in a separate PR.

If we make this stricter, the warning text / missing-sample estimate below should also stop assuming every irregular interval is a long gap.

```suggestion
  # (A) uniform-grid validation: any interval that differs from the expected
  # spacing indicates an irregular sample grid.
  irregular <- intervals != expected
```

**Review by mh105 — COMMENTED** [review]
Suggested change for strict uniform-grid detection.

File: R/utils-checks.R [hunk 3]
```diff
@@ -225,6 +225,234 @@ check_column <- function(df, col_name) {
...
+        "spacing (nominal rate: {nominal}). Estimated {n_missing} dropped ",
+        "sample(s); largest gap is {largest_gap} ms (~{largest_gap_ratio}x ",
+        "expected) before sample {first_gap_sample}. Some eye trackers drop ",
+        "samples instead of zero-filling missing pupil data, which violates ",
+        "the uniform-sampling assumption of the eyeris pipeline. Consider ",
+        "resampling onto a regular time grid before preprocessing."
+      )
+      log_warn(msg, verbose = verbose)
+    }
+  }
```
Comment (mh105) [inline R/utils-checks.R:460]: If `irregular` is changed to include both shorter and longer intervals, this block should stop assuming every irregular interval is a long gap. I think the warning should report both directions, while estimating dropped samples only from intervals longer than expected. The regularizer can handle the actual repair policy separately.

```suggestion
  if (n_irregular > 0) {
    long_intervals <- intervals[irregular & intervals > expected]
    short_intervals <- intervals[irregular & intervals < expected]
    n_long <- length(long_intervals)
    n_short <- length(short_intervals)

    # Estimate dropped samples only from intervals longer than expected.
    n_missing <- if (n_long > 0) {
      sum(pmax(round(long_intervals / expected) - 1, 0))
    } else {
      0
    }
    result$n_missing_samples <- as.integer(n_missing)

    if (verbose) {
      nominal <- if (is.finite(nominal_interval)) {
        paste0(hz, " Hz")
      } else {
        "inferred from data"
      }
      pct <- round(100 * result$prop_irregular, 2)
      first_irregular_sample <- which(irregular)[1] + 1
      long_gap_summary <- if (n_long > 0) {
        largest_gap <- round(max(long_intervals), 4)
        largest_gap_ratio <- round(max(long_intervals) / expected, 1)
        paste0(
          " Estimated ", n_missing, " dropped sample(s); largest long ",
          "interval is ", largest_gap, " ms (~", largest_gap_ratio,
          "x expected)."
        )
      } else {
        ""
      }

      msg <- paste0(
        "Non-uniform sampling intervals detected{segment}: {n_irregular} of ",
        "{n_total} intervals ({pct}%) differ from the expected ",
        "{expected_ms} ms spacing (nominal rate: {nominal}); ",
        "{n_long} longer and {n_short} shorter.{long_gap_summary} ",
        "First irregular interval occurs before sample ",
        "{first_irregular_sample}. This violates the uniform-sampling ",
        "assumption of the eyeris pipeline. Consider resampling onto a ",
        "regular time grid before preprocessing."
      )
      log_warn(msg, verbose = verbose)
    }
  }
```

**Review by mh105 — COMMENTED** [review]
Suggested follow-up for warning text and dropped-sample estimate.

File: tests/testthat/test-uniform-sampling.R [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+
+test_that("check_uniform_sampling_intervals counts multiple dropped samples", {
+  # two separate gaps: one 3 ms gap (2 missing) and one 4 ms gap (3 missing)
+  multi_gap <- c(0, 1, 2, 5, 6, 7, 11, 12)
+
+  res <- eyeris:::check_uniform_sampling_intervals(multi_gap, hz = 1000)
+  expect_false(res$uniform)
+  expect_equal(res$n_irregular, 2L)
+  expect_equal(res$n_missing_samples, 5L)
+})
```
Comment (mh105) [inline tests/testthat/test-uniform-sampling.R:67]: Can we add coverage for intervals that are shorter than the expected spacing? With the stricter uniform-grid definition, short intervals should be reported as irregular too, but they should not contribute to the dropped-sample estimate.

```suggestion
})

test_that("check_uniform_sampling_intervals detects short irregular intervals", {
  # 250 Hz: expected spacing is 4 ms. Shorter intervals still violate the
  # uniform sampling grid, even though they are not dropped-sample gaps.
  short_intervals <- c(0, 4, 7, 11, 13, 17)

  res <- eyeris:::check_uniform_sampling_intervals(short_intervals, hz = 250)
  expect_false(res$uniform)
  expect_equal(res$expected_interval, 4)
  expect_equal(res$n_irregular, 2L)
  expect_equal(res$n_missing_samples, 0L)

  warning_text <- captured_warning(eyeris:::check_uniform_sampling_intervals(
    short_intervals,
    hz = 250
  ))
  expect_match(warning_text, "Non-uniform sampling intervals detected")
  expect_match(warning_text, "0 longer")
  expect_match(warning_text, "2 shorter")
})
```

**Review by mh105 — COMMENTED** [review]
Suggested tests for short irregular intervals.

File: tests/testthat/test-uniform-sampling.R [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+  expect_equal(res_na$n_missing_samples, 1L)
+
+  # verbose = FALSE suppresses the warning message
+  msgs <- testthat::capture_messages(eyeris:::check_uniform_sampling_intervals(
+    c(0, 1, 2, 5, 6),
+    hz = 1000,
+    verbose = FALSE
+  ))
+  expect_length(msgs, 0)
+})
```
Comment (mh105) [inline tests/testthat/test-uniform-sampling.R:181]: Can we also add a regression test for the loud-failure path when the expected interval cannot be inferred? With finite positive intervals, `modal_value()` will always return some value, so this uses a non-finite timestamp difference as the realistic failure mode.

```suggestion
})

test_that("check_uniform_sampling_intervals errors when interval cannot be inferred", {
  expect_error(
    eyeris:::check_uniform_sampling_intervals(c(0, Inf), hz = 1000),
    "Unable to infer expected sampling interval"
  )
})
```

**Review by mh105 — COMMENTED** [review]
Suggested test for un-inferable expected interval.

File: R/utils-checks.R [hunk 6]
```diff
@@ -225,6 +225,234 @@ check_column <- function(df, col_name) {
...
+    return(invisible(result))
+  }
+
+  result$uniform <- FALSE
+  result$rate_mismatch <- rate_mismatch
+  result$n_irregular <- as.integer(n_irregular)
+  result$prop_irregular <- n_irregular / n_total
+
+  segment <- if (!is.null(block_label)) paste0(" in ", block_label) else ""
+  expected_ms <- round(expected, 4)
```
Comment (mh105) [inline R/utils-checks.R]: This value is only used for reporting in the warning, so four decimal places feels more precise than the data warrant. Since the timestamps are in milliseconds and common eye-tracking rates are around 250-1000 Hz, one decimal place should be enough for readable diagnostics while still covering sub-ms intervals such as 0.5 ms.

```suggestion
  expected_ms <- round(expected, 1)
```

**Review by mh105 — COMMENTED** [review]
Suggested reporting precision tweak.

**Review by mh105 — CHANGES_REQUESTED** [review]
